feat(ui): Add, Edit, Activate, Deactivate, Synchronize User from Admin - User Page

JaideepPalit opened the pull request:
> Add, Edit, Activate, Deactivate, Synchronize User from Admin -> User Page
    * Changed Clearing Request, Clearing Team field from free text field to user search dialog in create/edit Page
> * Which issue is this pull request belonging to and how is it solving it? (*#1242* , *#1241*)
> * Did you add or update any new dependencies that are required for your change? - No

Issues:
closes #1241 
closes #1242 

### How To Test?
   - Test Create/Edit Clearing Request for Clearing Team field
      - Condition to be able to edit CR Clearing Team - `isClearingTeam and not isClosedOrRejected` or `isRequestingUser`
   - Add, Edit, Activate, Deactivate, Synchronize User from Admin -> User Page
   - Synchronize User in both the DB, Scenarios
      - User present in Liferay DB, not present in CouchDB
      - User present in CouchDB, not present in Liferay DB
   - Deactivate
       -  If user is deactivated. User should not be able to login to 
           - sw360 from UI/Browser
           - Call REST API using API tokens, created from Preference tab
           - Call REST API using existing Bearer token/ OAuth Token. Token created before deactivating the user
           - Login to sw360-authorization server and create client details - http://localhost:8080/authorization/client-management 
           - Should not be able to refresh/create new token from pre-existing OAuth client details
           - Deactivated Users should not be listed when Search for User occurs like in Project, Component, Releases, Clearing Request - add moderators, contributors, Lead Architect, Security Responsible . etc
```
curl -X POST --user '94512*******:76cee06a-0161*************' -d 'grant_type=password&username=admin2@sw360.org&password=12345' http://localhost:8080/authorization/oauth/token -k

{
  "error" : "server_error",
  "error_description" : "403 "
}
```
   - To test on Stage or Sw360 with Single Sign-On Server
      - Set `sso.login.enabled=true` in `/etc/sw360/sw360.properties`
      - Test above mentioned steps.
      - Difference On Sw360 with Single Sign-On Server than in local
         - Password field is not present and cannot be set.
         - Try deleting a colleagues user from Liferay Control Panel, Regenerate the User from Admin -> User -> Add User.
         - Verify with your colleagues, if he is able to login, create, update, view his old projects.
> Have you implemented any additional tests?- No

### Checklist
Must:
- [X] All related issues are referenced in commit messages and in PR

Signed-off-by: Jaideep Palit <jaideep.palit@siemens.com>

JaideepPalit commented:
**Screenshots**
- Admin - User Page
![1](https://user-images.githubusercontent.com/56534710/120435931-f1435100-c39b-11eb-8280-c68094047d8c.JPG)

- Create New User
![2](https://user-images.githubusercontent.com/56534710/120435951-f4d6d800-c39b-11eb-981e-dc1340483b09.JPG)

- User Details Page
![6](https://user-images.githubusercontent.com/56534710/120435966-f7d1c880-c39b-11eb-8b25-ddfbcc7ef249.JPG)

- User Edit Page
![7](https://user-images.githubusercontent.com/56534710/120435975-fb654f80-c39b-11eb-842c-e1f2fe3091fd.JPG)

- Create clearing request 
![cr1](https://user-images.githubusercontent.com/56534710/120436267-57c86f00-c39c-11eb-8968-f9c7f8352c9b.JPG)

- Edit Clearing Request
![cr2](https://user-images.githubusercontent.com/56534710/120436275-59923280-c39c-11eb-89ea-1b3f0bebd92d.JPG)

JaideepPalit commented:
Resolved merge conflict

JaideepPalit commented:
Added Pagination for users page

akapti reviewed (changes requested):
Reviewed the code: suggested some indentation changes.

There are entries mismatch in `Language.properties` files, please fix it:
#### Screenshot 1:
![image](https://user-images.githubusercontent.com/50870174/124761051-012bf300-df4f-11eb-9fc4-23d51e77b3b4.png)

Two minor issues:

1. JavaScript error in console (screenshot below), on project list page, it is not impacting any functionality though. `Uncaught ReferenceError: require is not defined`

#### Screenshot 2:
![image](https://user-images.githubusercontent.com/50870174/124762208-4997e080-df50-11eb-8efe-99a28d91b5f1.png)

2. `Username` and `department` does not load in for `clearing team` search user modal pop-up, in edit mode for `clearing request`.

#### Screenshot 3:
![image](https://user-images.githubusercontent.com/50870174/124764699-de034280-df52-11eb-89b3-4b63b6200a4d.png)

JaideepPalit commented on frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/admin/UserPortlet.java:
```diff
@@ -77,97 +104,395 @@
-                String userGroup = "";
+                if (!externalIdFromReq.equals(userByEmailFromCouchDB.getExternalid())) {
+                    try {
+                        org.eclipse.sw360.datahandler.thrift.users.User newUser = userClient
```
Changed.
If user not found , it will return null from backend.
But thrift throws TException when null is transfered

JaideepPalit commented:
Thanks for review.

> There are entries mismatch in Language.properties files, please fix it:
>    - Updated file. Checked all newly added/deleted entries present in all 3 `language.properties` file

>JavaScript error in console (screenshot below), on project list page, it is not impacting any functionality though. Uncaught ReferenceError: require is not defined
>    - Fixed

> Username and department does not load in for clearing team search user modal pop-up, in edit mode for clearing request.
>    - There has been no changes in search Modal and functionality. It should behave similar to other places like adding moderator in project, release, component. However please let me know if you find any mismatch

akapti reviewed (commented):
Code looks good.
One more issue I found while testing is, If you create one `clearing request` from `Project list` page, then while creating second `clearing request`, the UI picks the previously selected `clearing team` user, Ideally it should not store the clearing team user for subsequent creation of clearing request in `project list` page..

JaideepPalit commented:
Thanks for review.
> One more issue I found while testing is, If you create one clearing request from Project list page, then while creating second clearing request, the UI picks the previously selected clearing team user, Ideally it should not store the clearing team user for subsequent creation of clearing request in project list page..
>  - Fixed

akapti reviewed (approved):
The above mentioned issue is fixed.
I have tested the `clearing request` functionality only.

@smrutis1  Please test the `User` related functionality in UI and also the REST API.

JaideepPalit commented:
Fixed issue related to department creation with orgMapper enabled.
`enable.custom.mapping=true` in sw360.properties

JaideepPalit commented:
Resolved merge conflict

smrutis1 commented:
Tested this PR in local instance:

1. Seems like the key for `secondary.department.and.roles` is missing in the language properties.

![image](https://user-images.githubusercontent.com/46560890/125619217-59cc6289-1b04-4bd7-aaa8-c0217f702adf.png)
2. May be some warning messages for the mandatory fields.

![image](https://user-images.githubusercontent.com/46560890/125620964-2477eadc-4b92-4094-9eab-44cd1c52ceb1.png)

3. Show `All` data gives empty, where as other counts are working.

![image](https://user-images.githubusercontent.com/46560890/125622271-b17f07b5-1414-4852-9a75-aa3c90d78d14.png)


4. When adding the user, if the email, gid is already present, it gives error. However the already entered form data vanishes, may be it would be good to retain it to avoid data entry again.

5. The count is not showing in the tooltip.
![Untitled](https://user-images.githubusercontent.com/46560890/125644294-a463367a-558c-4150-adfa-d24b3a06c721.png)

6. Sometimes getting this error [user_error.txt](https://github.com/eclipse/sw360/files/6827342/user_error.txt) in console(scenarios: if a user is only present in Liferay and user is trying to login for the first time.). May be a null check is necessary. 
7. May be add a row by default for the first time while clicking on the edit secondary roles and department icon from `Actions` column.
![image](https://user-images.githubusercontent.com/46560890/125883780-3779d2b6-73c3-465a-a340-0b736dd1c885.png)

8. After deactivating user, tried with both existing token "Token" and "Bearer" token, although both restricting access to the resources, but giving two different `error` message(`{
    "error": "unauthorized",
    "error_description": "User is deactivated"
}` & `{
    "error": "access_denied",
    "error_description": "User is deactivated"
}`) with response code `401` and `400` respectively. May be it should be consistent.

JaideepPalit commented:
Thanks for Test.
1. Fixed
2. (from discussion) red outline should indicate that it is mandatory field
3. Fixed
4. Fixed
5. Fixed
6. Fixed
7. Added
8. Fixed
